### DISCUSSION TITLE
Can one use both future and past covariates in sktime?

### DISCUSSION THREAD

**Discussion by atsangarides** [alan-turing-institute/sktime#1809] [Q&A]
Would like to make use of past (observed variables) and future covariates (forecasts that are available before a forecast is made for that horizon). Examples:
* past: observed temperature: can use lagged data as inputs to the model (i.e. temperature on day N-1 to predict sales on day N)
* future: forecasts known ahead of prediction (i.e. could be that we have the weather forecast for day N)

| Date      | weather_forecast_available | observed_temperature      | temperature_forecast | sales
| ----------- | ----------- |----------- |----------- |----------- |
2021-01-01	| 2020-12-31 |	35|	37|	100
2021-01-02	| 2021-01-01 |	28|	31 |	120

sktime abstracts the creation of lagged variables, etc, so any way to make the distinction between the two types of covariates? In the example above, `temperature_forecast` should not be lagged, as it is available ahead of prediction time if only doing day ahead sales forecasts. `observed_temperature`, however, should be lagged as the day needs to be completed to have the actual value. If both of these variables are used as exogenous in `X`, can we make the distinction in sktime?

**Comment by fkiraly** [comment]
apologies for the late reply, I've missed this over the holidays.

The `sktime` interface does not make a distinction, you need to put both in `X`.
Some forecasters will be able to use exogeneous variables that are available only up to the cutoff (current time), others will require that all variables in `X` are known for all indices to be forecast.

Regarding the lagging of different variables ("only lag the variables not available in the future), you can use the `ColumnTransformer`, which allows you to apply different transformers to different variables. You would have to address the variables by name or by column index though.

The current prototype (not merged yet) for the lagged window transformer can be applied to sets of variables separately:
https://github.com/alan-turing-institute/sktime/pull/1924

**Reply by atsangarides** [reply]
Thanks @fkiraly ! Really helpful!

Ok, one could make use of the `ColumnTransformer` as part of a `TransformedTargetForecaster` "pipeline". Cool, that's a nice way to go about it. 
`darts` implement this by defining which features are `past`/`future` covariates, as part of the forecaster itself, although the `ColumnTransformer` makes the lags more explicit (user is more aware of what is going on).

The lagged window transformer looks interesting, a bit similar to `tsfresh` I would think.
